[async_tp] Support ag+mm with gather_dim lastdim of mat_A

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #163068


Adding ag+mm support for the case, when gather_dim is last dim of matmul (reduction dim).

When we decompose matmul by reduction dimension we result in partials that needs additional reduction,
we allocate memory for accumulator.

Decomposition should not produce small (thin) mms that can not efficiently load the GPU. Limiting for minimal size of the shard 1024 (found empirically by testing in torchtitan).



scaled_mm is not supported yet for this case.


cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

IvanKobzarev committed:
[async_tp] Support ag+mm with gather_dim lastdim of mat_A

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[async_tp] Support ag+mm with gather_dim lastdim of mat_A"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[async_tp] Support ag+mm with gather_dim lastdim of mat_A"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[async_tp] Support ag+mm with gather_dim lastdim of mat_A"



Adding ag+mm support for the case, when gather_dim is last dim of matmul (reduction dim).

When we decompose matmul by reduction dimension we result in partials that needs additional reduction,
we allocate memory for accumulator.

This could be bad for performance and outweight the advantages of pipelining.

So we need to do more investigation when this decomposition can give performance wins.
Because of this I added a config that is turned off by default for experimentation and collecting results on different workloads.

scaled_mm was not tested for this decomposition and needs separate PR.




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[async_tp] Support ag+mm with gather_dim lastdim of mat_A"



Adding ag+mm support for the case, when gather_dim is last dim of matmul (reduction dim).

When we decompose matmul by reduction dimension we result in partials that needs additional reduction,
we allocate memory for accumulator.

This could be bad for performance and outweight the advantages of pipelining.

So we need to do more investigation when this decomposition can give performance wins.
Because of this I added a config that is turned off by default for experimentation and collecting results on different workloads.

scaled_mm was not tested for this decomposition and needs separate PR.




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[async_tp] Support ag+mm with gather_dim lastdim of mat_A"



Adding ag+mm support for the case, when gather_dim is last dim of matmul (reduction dim).

When we decompose matmul by reduction dimension we result in partials that needs additional reduction,
we allocate memory for accumulator.

This could be bad for performance and outweight the advantages of pipelining.

So we need to do more investigation when this decomposition can give performance wins.
Because of this I added a config that is turned off by default for experimentation and collecting results on different workloads.

scaled_mm was not tested for this decomposition and needs separate PR.




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[async_tp] Support ag+mm with gather_dim lastdim of mat_A"



Adding ag+mm support for the case, when gather_dim is last dim of matmul (reduction dim).

When we decompose matmul by reduction dimension we result in partials that needs additional reduction,
we allocate memory for accumulator.

Decomposition should not produce small (thin) mms that can not efficiently load the GPU. Limiting for minimal size of the shard 1024 (found empirically by testing in torchtitan).



scaled_mm is not supported yet for this case.


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

ngimel commented on torch/distributed/_symmetric_memory/__init__.py:
```diff
@@ -624,6 +637,143 @@ def default_consumer(shard: torch.Tensor, rank: int) -> None:
+    # shards[i] => shard from rank i
+    shards = []
+    for i, c in enumerate(ag_out.chunk(group_size)):
+        shards.append(c)
```
```suggestion
    shards = ag_out.chunk(group_size)
```

fegin commented on torch/_inductor/fx_passes/micro_pipeline_tp.py:
```diff
@@ -881,7 +904,7 @@ def fuse_matmul_reduce_scatter(reduce_scatter: _ReduceScatterMatch) -> None:
 
     filter_matmul = None
-    if orig_scatter_dim == _get_tensor(input_node).ndim - 1:
+    if _is_last_dim(_get_tensor(input_node), orig_scatter_dim):
```
Do you see the same small last dim len issue (< 1024) for fuse matmul reduce scatter?

IvanKobzarev replied:
Yes, for autoparallel torchtitan I observed that with TP dim = 8 and sequence length = 4096, the decomposition was slower. With sequence length = 8192 (chunked size = 1024) the AsyncTp decomposed was +4-5%.

IvanKobzarev committed:
Update on "[async_tp] Support ag+mm with gather_dim lastdim of mat_A"



Adding ag+mm support for the case, when gather_dim is last dim of matmul (reduction dim).

When we decompose matmul by reduction dimension we result in partials that needs additional reduction,
we allocate memory for accumulator.

Decomposition should not produce small (thin) mms that can not efficiently load the GPU. Limiting for minimal size of the shard 1024 (found empirically by testing in torchtitan).



scaled_mm is not supported yet for this case.


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

ngimel commented on torch/distributed/_symmetric_memory/__init__.py:
```diff
@@ -638,6 +788,15 @@ def _fused_all_gather_matmul_fallback(
         A_shard.contiguous(), group_size, group_name
     )
     A = torch.ops._c10d_functional.wait_tensor(A)
+    if gather_dim == A.ndim - 1 or gather_dim == -1:
```
```
import torch
allgathered = torch.empty(8, 18, 20)
group_size = 4
A_shard = torch.empty(2, 18, 20)
gather_dim = 2
A = allgathered
A = A.view(group_size, *A_shard.shape).movedim(gather_dim + 1, 1).flatten(0, 1)
print(A.movedim(0, gather_dim).shape) # generic path
A = allgathered
A_splits = A.chunk(group_size)
A_mm = torch.cat(A_splits, dim=-1)
print(A_mm.shape) # this path
```
prints the same shapes for both paths

IvanKobzarev commented:
@pytorchbot merge
